[10.0] l10n_ch_scan_bvr - Support of Postal BVR

Martronic-SA opened the pull request:
It's possible to have BVR for Postal accounts that are type 01 but with 53 positions.
Now your are getting the bvrnumber (bvr adherent number) and searching for it but they have no bvr adherent number, only the CCP.
By the way the bvrnumber for those BVR is not equal to '000000'.
Then if there is something in the bvrnumber, you should search for CCP/adherent = bvrnumber or CCP only (adherent = False)

Martronic-SA committed:
Support of Postal BVR

It's possible to have BVR for Postal accounts that are type 01 but with 53 positions.
Now your are getting the bvrnumber (bvr adherent number) and searching for it but they have no bvr adherent number, only the CCP.
By the way the bvrnumber for those BVR is not equal to '000000'.
Then if there is something in the bvrnumber, you should search for CCP/adherent = bvrnumber or CCP only (adherent = False)

yvaucher commented on l10n_ch_scan_bvr/wizard/scan_bvr.py:
```diff
@@ -353,8 +353,9 @@ def validate_bvr_string(self):
-                 ('bvr_adherent_num', '=', data['bvr_struct']['bvrnumber'])]
+                [('ccp', '=', data['bvr_struct']['beneficiaire']), '|',
+                 ('bvr_adherent_num', '=', data['bvr_struct']['bvrnumber']),
+                 ('bvr_adherent_num', '=', False)]
```
@Martronic-SA Shouldn't we refactor it to first search for a matching `bvr_adherent_num` then if not found search for a `bvr_adherent_num` set to false?

Plus, what if the ccp field is present on multiple accounts in your ERP? (with Bank's CCP)

You could end up matching the wrong supplier.

Martronic-SA replied:
This has been resolved by the commit https://github.com/OCA/l10n-switzerland/pull/447/commits/a8a8d316f5730525897a20a414169e23b738b6d9 below.

Martronic-SA committed:
Two steps search bank

Search first for postal account matching. If there is only one, we've found it, ok!
If there are many bank accounts with same ccp, check if we have a bvrnumber given (data['bvr_struct']['domain'] != 'name') and search for a ccp/bvr_adherent_num pair matching.

Martronic-SA committed:
Added space after # comment

yvaucher commented:
> It's possible to have BVR for Postal accounts that are type 01 but with 53 positions.

What do you mean by this? ISR ref are max 27 positions, and postal account number are 9 digits length (there was an old format with 6 but is obsolete).
Is `53` a typo?

Other than that I understand the module currently only works for ISR-B and not ISR for postal account.
And this patch should do the trick.

And side note this patch seems important as even if QR-Invoice will arrive soon, we can still receive ISR payment slips till 2022.

Martronic-SA commented:
> What do you mean by this? ISR ref are max 27 positions, and postal account number are 9 digits length (there was an old format with 6 but is obsolete).
> Is `53` a typo?

No in fact I wasn't speaking about the ISR ref itself, but the whole scanline thant can be 53 chars.

yvaucher commented:
@Martronic-SA Can you add a unit test with an example in
https://github.com/OCA/l10n-switzerland/blob/10.0/l10n_ch_scan_bvr/tests/test_scan_bvr.py ?

You can copy the test https://github.com/OCA/l10n-switzerland/blob/10.0/l10n_ch_scan_bvr/tests/test_scan_bvr.py#L73 and change the bvr_string by your case.

yvaucher commented:
set as `needs fixing` for additional unit test

Martronic-SA committed:
Update tests with 2 banks same CCP

Test scan wizard in the case of two bank accounts with same ccp but different bvr_adherent_num

Martronic-SA commented:
> 
> 
> set as `needs fixing` for additional unit test

I've updated the testings. I don't know where to check the coverage here but that should be ok.

yvaucher commented:
Indeed the travis configuration is missing coverage since a while. 

Last coverage was in nov 2019. I need to fix this.
https://coveralls.io/github/OCA/l10n-switzerland?branch=10.0

yvaucher reviewed (approved):
@Martronic-SA Thanks for the unit test.

Can you just squash your last commit please to keep the history cleaner?

Martronic-SA commented:
@yvaucher Is that ok now?

yvaucher commented:
/ocabot merge patch
